go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -212,18 +353,104 @@ func (npw *nodePortWatcher) SyncServices(services []interface{}) {
 				serviceInterface)
 			continue
 		}
-		npw.updateServiceFlowCache(service, true)
+		npw.updateServiceFlowCache(service, true, false)
 	}
 
 	npw.ofm.requestFlowSync()
-	syncSharedGatewayIptRules(services)
+	syncSharedGatewayIptRules(services, false, npw.nodeIPManager.addresses.List())
+}
+
+// Add new Local Traffic flows if endpoints are local
+func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
+	var serviceInfo interface{}
+	var exists bool
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+
+	svc, err := npw.watchFactory.GetService(ep.Namespace, ep.Name)
+	if err != nil {
+		// This is not necessarily an error. For e.g when there are endpoints
+		// without a corresponding service.
+		klog.V(5).Infof("No service found for endpoint %s in namespace %s", ep.Name, ep.Namespace)
+		return
+	}
+
+	if !util.ServiceTypeHasClusterIP(svc) || !util.IsClusterIPSet(svc) {
+		return
+	}
+
+	// Endpoint add Event can arrive before service add event, see if service exists
+	// create it if it doesn't exist, store endpoint if it does
+	if serviceInfo, exists = npw.services.LoadOrStore(name, svcInfo{service: svc, endpoints: ep}); !exists {
+		klog.V(5).Infof("Endpoint: %s event in namespace: %s arrived before service creation, create service: %s",
+			ep.Name, ep.Namespace, ep.Name)
+		npw.AddService(svc)
+	} else {
+		npw.services.Store(name, svcInfo{service: serviceInfo.(svcInfo).service, endpoints: ep})
+	}
+
+	if hasHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses) && util.ServiceExternalTrafficPolicyLocal(serviceInfo.(svcInfo).service) {
+		klog.V(5).Infof("Service %s has ExternalTrafficPolicy=local and local Endpoints, Updating flow Cache", ep.Name)
+		// Update Service Cache with special flows
+		npw.updateServiceFlowCache(serviceInfo.(svcInfo).service, true, true)
+		npw.ofm.requestFlowSync()
+		delSharedGatewayIptRules(serviceInfo.(svcInfo).service, false, npw.nodeIPManager.addresses.List())
+		addSharedGatewayIptRules(serviceInfo.(svcInfo).service, true, npw.nodeIPManager.addresses.List())
+	}
+}
+
+func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+	if serviceInfo, exists := npw.services.LoadAndDelete(name); exists && util.ServiceExternalTrafficPolicyLocal(serviceInfo.(svcInfo).service) &&
+		hasHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses) {
+
+		npw.updateServiceFlowCache(serviceInfo.(svcInfo).service, false, true)
+		npw.ofm.requestFlowSync()
+		delSharedGatewayIptRules(serviceInfo.(svcInfo).service, true, npw.nodeIPManager.addresses.List())
+
+	}
+	npw.services.Delete(name)
+}
+
+func (npw *nodePortWatcher) UpdateEndpoints(old *kapi.Endpoints, new *kapi.Endpoints) {
```

Comment:
update is getting tricky, the more im looking at this the more im thinking we are going to need a mutex on the service itself to avoid stepping on each other. I might come up with a commit to show you what I'm thinking, its hard to explain over comments.

Reply:
I think its fine doing it this way, especially because the dependency is one directional 

- If a service update happens we always check to see if we missed any endpoint updates and manage flows
- If an endpoint event happens we just manage flows 
      -  If a svc event arrives during an endpoint event, the first endpoint event will finish, and then the Service 
         Update event will trigger another endpoint event to clean anything up that happened 

Now currently this is broken for the case that we had hostnetworked and now do not because it wouldn't correctly add back the old flows but otherwise I can't think of any other holes at the moment